core/lib/contract_verifier/src/resolver/github/mod.rs

```diff
@@ -0,0 +1,311 @@
+use std::{
+    collections::HashMap,
+    time::{Duration, Instant},
+};
+
+use anyhow::Context as _;
+use tokio::{io::AsyncWriteExt as _, sync::RwLock};
+use zksync_queued_job_processor::async_trait;
+
+use self::gh_api::GitHubApi;
+use crate::{
+    compilers::{Solc, SolcInput, Vyper, VyperInput, ZkSolc, ZkSolcInput, ZkVyper},
+    error::ContractVerifierError,
+    resolver::{
+        Compiler, CompilerPaths, CompilerResolver, CompilerType, SupportedCompilerVersions,
+    },
+    ZkCompilerVersions,
+};
+
+mod gh_api;
+
+/// [`CompilerResolver`] that can dynamically download missing compilers from GitHub releases.
+///
+/// Note: this resolver does not interact with [`EnvCompilerResolver`](super::EnvCompilerResolver).
+/// This is important for the context of zksolc/zkvyper, as there we have two separate compilers
+/// required for compilation. This resolver will download both of them, even if one of the versions
+/// is available in the `EnvCompilerResolver`.
+#[derive(Debug)]
+pub(crate) struct GitHubCompilerResolver {
+    /// We expect that contract-verifier will be running in docker without any persistent storage,
+    /// so we explicitly don't expect any artifacts to survive restart.
+    artifacts_dir: tempfile::TempDir,
+    gh_client: GitHubApi,
+    client: reqwest::Client,
+    supported_versions: RwLock<SupportedVersions>,
+    /// List of downloads performed right now.
+    /// `broadcast` receiver can be used to wait until the download is finished.
+    active_downloads: RwLock<HashMap<(CompilerType, String), tokio::sync::broadcast::Receiver<()>>>,
```

Comment:
Nit: I think `watch` channels would suffice here; there seems to be a single producer. Not a major concern; can be addressed in a follow-up PR.